### PR TITLE
Added 29bit ID support to cansniffer

### PR DESCRIPTION
**Pull request by coryjfowler**
Moved from CAN_BCM to CAN_RAW.  I have most of the functionality restored.  I just need to look into how CAN_RAW filtering works and add that ability.

### CODE REVIEW

File: cansniffer.c [hunk 1]
```diff
@@ -60,7 +61,7 @@
 #include <net/if.h>
 
 #include <linux/can.h>
-#include <linux/can/bcm.h>
+//#include <linux/can/bcm.h>
```
Comment (marckleinebudde) [inline cansniffer.c]: Please remove and do not comment out.

File: cansniffer.c [hunk 2]
```diff
@@ -119,15 +120,28 @@ static unsigned char binary_gap;
 static unsigned char color;
 static char *interface;
 
-void rx_setup (int fd, int id);
-void rx_delete (int fd, int id);
-void print_snifline(int id);
+//void rx_setup (int fd, unsigned long id);
+//void rx_delete (int fd, unsigned long id);
```
Comment (marckleinebudde) [inline cansniffer.c]: same here

File: cansniffer.c [hunk 3]
```diff
@@ -119,15 +120,28 @@ static unsigned char binary_gap;
...
+void print_snifline(int index);
 int handle_keyb(int fd);
-int handle_bcm(int fd, long currcms);
+int handle_frame(int fd, long currcms);
 int handle_timeo(int fd, long currcms);
 void writesettings(char* name);
 void readsettings(char* name, int sockfd);
 
+int idx = 0;
```
Comment (marckleinebudde) [inline cansniffer.c]: static int idx;

Comment (marckleinebudde) [reply cansniffer.c]: as it's a global it will be initialized to 0 automatically.

File: cansniffer.c [hunk 4]
```diff
@@ -119,15 +120,28 @@ static unsigned char binary_gap;
...
+
+
+int comp (const void * elem1, const void * elem2) 
+{
+    unsigned long f = ((struct snif*)elem1)->current.can_id;
+    unsigned long s = ((struct snif*)elem2)->current.can_id;
+    if (f > s) return  1;
```
Comment (marckleinebudde) [inline cansniffer.c]: don't put if and return in same line

File: cansniffer.c [hunk 5]
```diff
@@ -119,15 +120,28 @@ static unsigned char binary_gap;
...
+    if (f < s) return -1;
+    return 0;
+}
+
+
```
Comment (marckleinebudde) [inline cansniffer.c]: one newline is enough

File: cansniffer.c [hunk 6]
```diff
@@ -302,14 +316,23 @@ int main(int argc, char **argv)
...
+//			rx_setup(s, i);
+	
+//	for (unsigned long temp = 0x80000000; temp < 0xA0000000; temp++)
+//		rx_setup(s, 0x89F801CF);
+//		rx_setup(s, 0x89F802CF);
+//		rx_setup(s, 0x8DF010CF);
+//		rx_setup(s, 0x8DF805CF);
+//		rx_setup(s, 0x99FA03CF);
+//		rx_setup(s, 0x99FA04CF);
+
```
Comment (marckleinebudde) [inline cansniffer.c]: please remove, do not comment out

File: cansniffer.c [hunk 7]
```diff
@@ -338,57 +361,57 @@ int main(int argc, char **argv)
...
+			running &= handle_frame(s, currcms);
 
 		if (currcms - lastcms >= loop) {
 			running &= handle_timeo(s, currcms);
 			lastcms = currcms;
 		}
 	}
 
-	printf("%s", CSR_SHOW); /* show cursor */
+	printf("%s\n", CSR_SHOW); /* show cursor */
```
Comment (marckleinebudde) [inline cansniffer.c]: Seems to be an unrelated change, please move into a separate commit.

File: cansniffer.c [hunk 8]
```diff
@@ -338,57 +361,57 @@ int main(int argc, char **argv)
...
+//
+//	struct bcm_msg_head msg_head;
+//
+//	msg_head.opcode  = RX_DELETE;
+//	msg_head.can_id  = id;
+//	msg_head.nframes = 0;
+//
+//	if (write(fd, &msg_head, sizeof(msg_head)) < 0)
+//		perror("write");
+//}
```
Comment (marckleinebudde) [inline cansniffer.c]: please remove

File: cansniffer.c [hunk 9]
```diff
@@ -419,15 +442,15 @@ int handle_keyb(int fd){
 			for (i=0; i < 2048 ;i++) {
 				if (((i & mask) == (value & mask)) && (is_clr(i, ENABLE))) {
 					do_set(i, ENABLE);
-					rx_setup(fd, i);
+					//rx_setup(fd, i);
```
Comment (marckleinebudde) [inline cansniffer.c]: same here

File: cansniffer.c [hunk 10]
```diff
@@ -419,15 +442,15 @@ int handle_keyb(int fd){
...
 				}
 			}
 		}
 		else { /* '-' */
 			for (i=0; i < 2048 ;i++) {
 				if (((i & mask) == (value & mask)) && (is_set(i, ENABLE))) {
 					do_clr(i, ENABLE);
-					rx_delete(fd, i);
+					//rx_delete(fd, i);
```
Comment (marckleinebudde) [inline cansniffer.c]: same here

File: cansniffer.c [hunk 11]
```diff
@@ -489,45 +512,78 @@ int handle_keyb(int fd){
...
-	int nbytes, id, i;
+	unsigned long id;
 
-	struct {
-		struct bcm_msg_head msg_head;
+//	struct {
+//		struct bcm_msg_head msg_head;
 		struct can_frame frame;
-	} bmsg;
+//	} bmsg;
```
Comment (marckleinebudde) [inline cansniffer.c]: same here

File: cansniffer.c [hunk 12]
```diff
@@ -489,45 +512,78 @@ int handle_keyb(int fd){
...
-	id = bmsg.msg_head.can_id;
-	ioctl(fd, SIOCGSTAMP, &sniftab[id].currstamp);
+	id = frame.can_id;
 
-	if (bmsg.msg_head.opcode != RX_CHANGED) {
-		printf("received strange BCM opcode %d!\n", bmsg.msg_head.opcode);
-		return 0; /* quit */
+	for(i = 0; i <= idx; i++){
+		if(id == sniftab[i].current.can_id){
+			pos = i;
```
Comment (marckleinebudde) [inline cansniffer.c]: space after for and if

File: cansniffer.c [hunk 13]
```diff
@@ -489,45 +512,78 @@ int handle_keyb(int fd){
...
+			match = true;
+			break;
+		}
 	}
 
-	if (nbytes != sizeof(bmsg)) {
-		printf("received strange BCM data length %d!\n", nbytes);
+	if(!match){
+		pos = idx++;
```
Comment (marckleinebudde) [inline cansniffer.c]: same here

File: cansniffer.c [hunk 14]
```diff
@@ -489,45 +512,78 @@ int handle_keyb(int fd){
...
+			}
+		else
+			rx_changed = true;
+	}
+
+//	if (bmsg.msg_head.opcode != RX_CHANGED) {
+//		printf("received strange BCM opcode %d!\n", bmsg.msg_head.opcode);
+//		return 0; /* quit */
+//	}
+
```
Comment (marckleinebudde) [inline cansniffer.c]: please remove

File: cansniffer.c (hunk 13 above)
Comment (marckleinebudde) [reply cansniffer.c]: You probably want to write it:
```
if () {
} else if () {
} else {
}
```

File: cansniffer.c [hunk 15]
```diff
@@ -489,45 +512,78 @@ int handle_keyb(int fd){
...
 
-	sniftab[id].current = bmsg.frame;
-	for (i=0; i < 8; i++)
-		sniftab[id].marker.data[i] |= sniftab[id].current.data[i] ^ sniftab[id].last.data[i];
+	if(rx_changed == true){
+		ioctl(fd, SIOCGSTAMP, &sniftab[pos].currstamp);
 
-	sniftab[id].timeout = (timeout)?(currcms + timeout):0;
+		sniftab[pos].current = frame;
+		for (i=0; i < 8; i++)
```
Comment (marckleinebudde) [inline cansniffer.c]:
```
for (i = 0; i < 8; i++)

File: cansniffer.c [hunk 16]
```diff
@@ -593,11 +649,12 @@ int handle_timeo(int fd, long currcms){
 
 };
 
-void print_snifline(int id){
+void print_snifline(int position){
 
-	long diffsec  = sniftab[id].currstamp.tv_sec  - sniftab[id].laststamp.tv_sec;
-	long diffusec = sniftab[id].currstamp.tv_usec - sniftab[id].laststamp.tv_usec;
-	int dlc_diff  = sniftab[id].last.can_dlc - sniftab[id].current.can_dlc;
+	long diffsec  = sniftab[position].currstamp.tv_sec  - sniftab[position].laststamp.tv_sec;
```
Comment (marckleinebudde) [inline cansniffer.c]: please leave only one space before and after the `-`. I know you didn't add this... :)

File: cansniffer.c [hunk 17]
```diff
@@ -609,20 +666,23 @@ void print_snifline(int id){
 	if (diffsec > 10)
 		diffsec = 9, diffusec = 999999;
 
-	printf("%ld.%06ld  %3X  ", diffsec, diffusec, id);
+	if((id & 0x80000000) == 0x80000000)
```
Comment (marckleinebudde) [inline cansniffer.c]: space after if

File: cansniffer.c [hunk 18]
```diff
@@ -609,20 +666,23 @@ void print_snifline(int id){
...
+		printf("%ld.%06ld EXT 0x%08lX  ", diffsec, diffusec, (id & 0x1FFFFFFF));
+	else
+		printf("%ld.%06ld STD 0x%03lX       ", diffsec, diffusec, id);
 
 	if (binary) {
 
-		for (i=0; i<sniftab[id].current.can_dlc; i++) {
+		for (i=0; i<sniftab[position].current.can_dlc; i++) {
```
Comment (marckleinebudde) [inline cansniffer.c]: Please fix the spaces here, too.

File: cansniffer.c [hunk 19]
```diff
@@ -643,22 +703,22 @@ void print_snifline(int id){
 	}
 	else {
 
-		for (i=0; i<sniftab[id].current.can_dlc; i++)
-			if ((color) && (sniftab[id].marker.data[i]) && (!(sniftab[id].notch.data[i])))
-				printf("%s%02X%s ", ATTCOLOR, sniftab[id].current.data[i], ATTRESET);
+		for (i=0; i<sniftab[position].current.can_dlc; i++)
```
Comment (marckleinebudde) [inline cansniffer.c]: and here

File: cansniffer.c [hunk 20]
```diff
@@ -643,22 +703,22 @@ void print_snifline(int id){
...
-			printf("%*s", (8 - sniftab[id].current.can_dlc) * 3, "");
+		if (sniftab[position].current.can_dlc < 8)
+			printf("%*s", (8 - sniftab[position].current.can_dlc) * 3, "");
 
-		for (i=0; i<sniftab[id].current.can_dlc; i++)
-			if ((sniftab[id].current.data[i] > 0x1F) && 
-			    (sniftab[id].current.data[i] < 0x7F))
-				if ((color) && (sniftab[id].marker.data[i]) && (!(sniftab[id].notch.data[i])))
-					printf("%s%c%s", ATTCOLOR, sniftab[id].current.data[i], ATTRESET);
+		for (i=0; i<sniftab[position].current.can_dlc; i++)
```
Comment (marckleinebudde) [inline cansniffer.c]: and here

File: cansniffer.c [hunk 21]
```diff
@@ -723,15 +783,15 @@ void readsettings(char* name, int sockfd){
 				if (buf[5] & 1) {
 					if (is_clr(i, ENABLE)) {
 						do_set(i, ENABLE);
-						if (sockfd)
-							rx_setup(sockfd, i);
+//						if (sockfd)
+//							rx_setup(sockfd, i);
```
Comment (marckleinebudde) [inline cansniffer.c]: please remove

File: cansniffer.c [hunk 22]
```diff
@@ -723,15 +783,15 @@ void readsettings(char* name, int sockfd){
...
 					}
 				}
 				else
 					if (is_set(i, ENABLE)) {
 						do_clr(i, ENABLE);
-						if (sockfd)
-							rx_delete(sockfd, i);
+//						if (sockfd)
+//							rx_delete(sockfd, i);
```
Comment (marckleinebudde) [inline cansniffer.c]: please remove

File: cansniffer.c [hunk 23]
```diff
@@ -814,3 +764,14 @@ void readsettings(char* name, int sockfd){
 	else
 		printf("unable to read setting file '%s'!\n", fname);
 };
+
+int sniftab_index(canid_t id){
```
Comment (marckleinebudde) [inline cansniffer.c]: static
and a space between `) {`

File: README.md [hunk 24]
```diff
@@ -13,7 +13,7 @@ subsystem (aka SocketCAN):
 * canplayer : replay CAN logfiles
 * cansend : send a single frame
 * cangen : generate (random) CAN traffic
-* cansniffer : display CAN data content differences (just 11bit CAN IDs)
+* cansniffer : display CAN data content differences (~~just 11bit CAN IDs~~ Not any more!)
```
Comment (marckleinebudde) [inline README.md]: Just remove the comment in the about the 11 bit. We've git to have a history of the source code.

**Comment by coryjfowler** [comment]
I hope that works...  I'm not very fluent with 'git'.

File: cansniffer.c (hunk 15 above)
Comment (coryjfowler) [reply cansniffer.c]: The original code has a lot of suppressed whitespace in the existing for loops.

File: cansniffer.c [hunk 25]
```diff
@@ -200,21 +216,21 @@ int main(int argc, char **argv)
 	signal(SIGHUP, sigterm);
 	signal(SIGINT, sigterm);
 
-	for (i=0; i < 2048 ;i++) /* default: check all CAN-IDs */
+	for (i=0; i < MAX_SLOTS ;i++) /* default: enable all slots */
```
Comment (marckleinebudde) [inline cansniffer.c]: ``for (i=0; i < MAX_SLOTS; i++)``

File: cansniffer.c [hunk 26]
```diff
@@ -489,45 +451,68 @@ int handle_keyb(int fd){
...
+		match = true;
+
+	if(!match){
+		if(idx < MAX_SLOTS){
+			pos = idx++;
+			rx_changed = true;
+		}
+	}
+	else {
+		if(frame.can_dlc == sniftab[pos].current.can_dlc)
```
Comment (marckleinebudde) [inline cansniffer.c]: ``if (``

File: cansniffer.c [hunk 27]
```diff
@@ -489,45 +451,68 @@ int handle_keyb(int fd){
...
+			for(i=0; i<frame.can_dlc; i++){
```
Comment (marckleinebudde) [inline cansniffer.c]: ``for (i = 0; i < frame.can_dlc; i++) {``

File: cansniffer.c [hunk 28]
```diff
@@ -489,45 +451,68 @@ int handle_keyb(int fd){
...
+				if(frame.data[i] != sniftab[pos].current.data[i] ){
```
Comment (marckleinebudde) [inline cansniffer.c]: ``if (frame.data[i] != sniftab[pos].current.data[i]) {``

File: cansniffer.c [hunk 29]
```diff
@@ -489,45 +451,68 @@ int handle_keyb(int fd){
...
 
-	sniftab[id].current = bmsg.frame;
-	for (i=0; i < 8; i++)
-		sniftab[id].marker.data[i] |= sniftab[id].current.data[i] ^ sniftab[id].last.data[i];
+	if(rx_changed == true){
+		ioctl(fd, SIOCGSTAMP, &sniftab[pos].currstamp);
 
-	sniftab[id].timeout = (timeout)?(currcms + timeout):0;
+		sniftab[pos].current = frame;
+		for (i=0; i < 8; i++)
```
Comment (marckleinebudde) [inline cansniffer.c]: ``for (i = 0; i < 8; i++)``

File: cansniffer.c [hunk 30]
```diff
@@ -643,22 +632,22 @@ void print_snifline(int id){
 	}
 	else {
 
-		for (i=0; i<sniftab[id].current.can_dlc; i++)
-			if ((color) && (sniftab[id].marker.data[i]) && (!(sniftab[id].notch.data[i])))
-				printf("%s%02X%s ", ATTCOLOR, sniftab[id].current.data[i], ATTRESET);
+		for (i=0; i<sniftab[slot].current.can_dlc; i++)
```
Comment (marckleinebudde) [inline cansniffer.c]: ``for (i = 0; i < sniftab[slot].current.can_dlc; i++)``

File: cansniffer.c [hunk 31]
```diff
@@ -643,22 +632,22 @@ void print_snifline(int id){
...
-			printf("%*s", (8 - sniftab[id].current.can_dlc) * 3, "");
+		if (sniftab[slot].current.can_dlc < 8)
+			printf("%*s", (8 - sniftab[slot].current.can_dlc) * 3, "");
 
-		for (i=0; i<sniftab[id].current.can_dlc; i++)
-			if ((sniftab[id].current.data[i] > 0x1F) && 
-			    (sniftab[id].current.data[i] < 0x7F))
-				if ((color) && (sniftab[id].marker.data[i]) && (!(sniftab[id].notch.data[i])))
-					printf("%s%c%s", ATTCOLOR, sniftab[id].current.data[i], ATTRESET);
+		for (i=0; i<sniftab[slot].current.can_dlc; i++)
```
Comment (marckleinebudde) [inline cansniffer.c]: ``for (i = 0; i < sniftab[slot].current.can_dlc; i++)``

File: cansniffer.c [hunk 32]
```diff
@@ -684,19 +672,18 @@ void writesettings(char* name){
...
 	strncat(fname, name, 29 - strlen(fname)); 
-	fd = open(fname,  O_WRONLY|O_CREAT, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
+	fd = open(fname,  O_WRONLY|O_CREAT|O_TRUNC, S_IRUSR|S_IWUSR|S_IRGRP|S_IROTH);
     
 	if (fd > 0) {
-
-		for (i=0; i < 2048 ;i++) {
-			sprintf(buf, "<%03X>%c.", i, (is_set(i, ENABLE))?'1':'0');
-			write(fd, buf, 7);
+		for (i=0; i < idx; i++) {
```
Comment (marckleinebudde) [inline cansniffer.c]: ``for (i = 0; i < idx; i++) {``

File: cansniffer.c [hunk 33]
```diff
@@ -684,19 +672,18 @@ void writesettings(char* name){
...
+			sprintf(buf, "<%08X>%c.", sniftab[i].current.can_id, (is_set(i, ENABLE))?'1':'0');
+			write(fd, buf, 12);
 			for (j=0; j<8 ; j++){
```
Comment (marckleinebudde) [inline cansniffer.c]: While there pleas also fix:
``for (j = 0;  j < 8 ; j++){``

File: cansniffer.c [hunk 34]
```diff
@@ -709,40 +696,50 @@ void readsettings(char* name, int sockfd){
...
-
-		for (i=0; i < 2048 ;i++) {
-			if (read(fd, &buf, 24) == 24) {
-				if (buf[5] & 1) {
-					if (is_clr(i, ENABLE)) {
-						do_set(i, ENABLE);
-						if (sockfd)
-							rx_setup(sockfd, i);
+ 
+		while(!done){
```
Comment (marckleinebudde) [inline cansniffer.c]: ``while (!done) {``

File: cansniffer.c [hunk 35]
```diff
@@ -709,40 +696,50 @@ void readsettings(char* name, int sockfd){
...
+			if (read(fd, &buf, 29) == 29) {
+				unsigned long id = strtoul(&buf[1], (char **)NULL, 16);
+				pos = sniftab_index(id);
+
+				if(pos < 0){
```
Comment (marckleinebudde) [inline cansniffer.c]: ``if (pos < 0) {``

File: cansniffer.c [hunk 36]
```diff
@@ -709,40 +696,50 @@ void readsettings(char* name, int sockfd){
...
+					if(idx < MAX_SLOTS){
```
Comment (marckleinebudde) [inline cansniffer.c]: ``if (idx < MAX_SLOTS) {``

File: cansniffer.c [hunk 37]
```diff
@@ -709,40 +696,50 @@ void readsettings(char* name, int sockfd){
...
+						pos = idx++;
+						sniftab[pos].current.can_id = id;
+					}
+					else
+						break;
+				}
+				if (buf[10] & 1) {
+					if (is_clr(pos, ENABLE)) {
+						do_set(pos, ENABLE);
```
Comment (marckleinebudde) [inline cansniffer.c]: ``if (buf[10] & 1 && is_clr(pos, ENABLE)) {``

File: cansniffer.c [hunk 38]
```diff
@@ -709,40 +696,50 @@ void readsettings(char* name, int sockfd){
...
 				}
 				else
-					if (is_set(i, ENABLE)) {
-						do_clr(i, ENABLE);
-						if (sockfd)
-							rx_delete(sockfd, i);
+					if (is_set(pos, ENABLE)) {
+						do_clr(pos, ENABLE);
 					}
 				for (j=7; j>=0 ; j--){
```
Comment (marckleinebudde) [inline cansniffer.c]: ``for (j = 7;  j >= 0; j--) {``

File: cansniffer.c [hunk 39]
```diff
@@ -709,40 +696,50 @@ void readsettings(char* name, int sockfd){
...
-					sniftab[i].notch.data[j] =
-						(__u8) strtoul(&buf[2*j+7], (char **)NULL, 16) & 0xFF;
-					buf[2*j+7] = 0; /* cut off each time */
+					sniftab[pos].notch.data[j] =
+						(__u8) strtoul(&buf[2*j+12], (char **)NULL, 16) & 0xFF;
+					buf[2*j+12] = 0; /* cut off each time */
```
Comment (marckleinebudde) [inline cansniffer.c]: ``buf[2 * j + 12] = 0``

File: cansniffer.c [hunk 40]
```diff
@@ -709,40 +696,50 @@ void readsettings(char* name, int sockfd){
...
-							rx_delete(sockfd, i);
+					if (is_set(pos, ENABLE)) {
+						do_clr(pos, ENABLE);
 					}
 				for (j=7; j>=0 ; j--){
-					sniftab[i].notch.data[j] =
-						(__u8) strtoul(&buf[2*j+7], (char **)NULL, 16) & 0xFF;
-					buf[2*j+7] = 0; /* cut off each time */
+					sniftab[pos].notch.data[j] =
+						(__u8) strtoul(&buf[2*j+12], (char **)NULL, 16) & 0xFF;
```
Comment (marckleinebudde) [inline cansniffer.c]: You can drop the casts and the ``& 0xff``.
``strtoul(&buf[2 * j + 12], NULL, 16);``

File: cansniffer.c [hunk 41]
```diff
@@ -754,3 +751,14 @@ void readsettings(char* name, int sockfd){
 	else
 		printf("unable to read setting file '%s'!\n", fname);
 };
+
+int sniftab_index(canid_t id){
+
+	int i;
+
+	for(i = 0; i <= idx; i++)
+		if(id == sniftab[i].current.can_id)
```
Comment (marckleinebudde) [inline cansniffer.c]: ``if (id == sniftab[i].current.can_id)``

**Comment by hartkopp** [comment]
Dear @marckleinebudde and @coryjfowler , 
I've just followed your discussion and I'm personally pretty unsure, how to proceed with cansniffer.
The original approach was to use the entire CAN_BCM functionality to just display the differences filtered inside BCM socket (and therefore in kernel space).
With 29bit IDs this becomes unusable as it makes no sense to create 2^29 BCM-jobs in a socket in the kernel, right? Therefore I tried to merge some portions of candump (using CAN_RAW) and the output generation of cansniffer into a new implementation here: https://github.com/hartkopp/can-utils/commits/master/cansniffer.c 
My idea was to support more than one CAN bus at a time, provide a different colorization for them, support CAN FD and sort the output by CAN interface OR by CAN ID.

Maybe my goals where just to ambitious that I got stuck with the implementation?!?

But we need a complete new concept to handle, sort (by ID) and compare incoming frames from CAN_RAW - and I wonder if using the CAN_RAW filters makes sense at all :-/

Any thoughts?

File: cansniffer.c (hunk 37 above)
Comment (coryjfowler) [reply cansniffer.c]: I feel that this would create undesired behavior unless I modify the 'else' that follows it.

**Comment by coryjfowler** [comment]
Hello @hartkopp,

Right now this implementation checks to see if an ID exists in the sniftab array.  It will then either create a new slot or check the existing slot for changes in the frame.  If there are changes, it updates the struct in that slot like the original code.  It will then sort the sniftab array by current.ID using 'qsort' so that it is displayed nicely in order with a simple for loop.

My aim for socket filtering is to help alleviate the program from handling a bus with potentially more than 2048 IDs in active use, but the direction here could also just be to change the value of MAX_SLOTS to something sane and recompile.

**Comment by SSinSD** [comment]
I am very interested in using cansniffer for some reverse engineering on my Chevy SS and the GMLAN uses 29 bit extended IDs.  Any chance this update to support extended IDs will be merged into the master anytime soon?

**Comment by brandonros** [comment]
This isn't clear: does candump support 29-bit messages as it stands now? Not cansniffer, but candump?

**Comment by hartkopp** [comment]
candump can deal with all kinds for CAN traffic (11bit / 29bit Identifiers, Classic CAN and CAN FD). cansniffer can only cope with 11bit IDs and Classic CAN (8 bytes payload).

**Comment by mitchdetailed** [comment]
The method I used to get around this is to set up a vcan network and mask the id to the 11b value, and send it onto vcan0.

Example shown here: https://github.com/mitchdetailed/python-can-re/blob/master/can0ext-vcan1masked.py

**Comment by peteg944** [comment]
Still no progress? It's been a year and a half... RIP. Are there any alternatives to cansniffer that work with the full-length IDs?

**Comment by coryjfowler** [comment]
> 
> 
> Still no progress? It's been a year and a half... RIP. Are there any alternatives to cansniffer that work with the full-length IDs?

The version on my github has the functionality.  However, it has not been accepted here because of styling issues that existed before I even made changes and I simply do not have the time to update those.  So its being held back by purely petty reasons...

**Comment by peteg944** [comment]
> > Still no progress? It's been a year and a half... RIP. Are there any alternatives to cansniffer that work with the full-length IDs?
> 
> The version on my github has the functionality. However, it has not been accepted here because of styling issues that existed before I even made changes and I simply do not have the time to update those. So its being held back by purely petty reasons...

Sorry about that, I didn't mean anything towards you. My sarcasm was toward the people who maintain these "can-utils", lol. Thank you for taking the time to actually think of a solution and write the code in the way that it should've been written in the first place! I read through all the comments and like you said, it was all for stylistic reasons which don't really matter.

**Comment by hartkopp** [comment]
Would you like to test some 29bit cansniffer rework?
See https://github.com/linux-can/can-utils/issues/181

BR,
Oliver

**Comment by hartkopp** [comment]
Followed up in issue https://github.com/linux-can/can-utils/issues/181
Integrated this suggested slot approach from @coryjfowler with some more fixes and improvements.
@coryjfowler @SSinSD @brandonros @mitchdetailed @peteg944 : 
You are invited to test (and review) this pull request: https://github.com/linux-can/can-utils/pull/200
The latest code can be found here: https://github.com/hartkopp/can-utils/tree/29bit-cansniffer
The single steps of development and improvements can be found here: https://github.com/hartkopp/can-utils/tree/snif
